### PR TITLE
Add script to post process file::unpack

### PR DESCRIPTION
**Pull request by coolo**
Please review - to be called from Cavil::Checkout

### CODE REVIEW

**Comment by coolo** [comment]
thinking about it - an extra script does most likely no sense and I should integrate this into Checkout directly

preferences @kraih ?

**Comment by kraih** [comment]
Yes, it should definitely be integrated. Not sure how i feel about mangling the caches source files though.

**Comment by kraih** [comment]
To be more precise, this should probably be a utility class used by `Checkout.pm`. So it can be tested standalone.

**Comment by coolo** [comment]
We can also translate .unpacked.json to .postprocessed.json

**Comment by coolo** [comment]
but actually we need to postprocess even more - because file::unpack isn't mean to process the files.

**Comment by coolo** [comment]
I guess I'll stop here - but my inner me also wants to move the html processing into mojo and out of file::unpack.

Because if you check how many files are mis-sniffed as html and forced into w3m, it makes you cry. At this point we might be better off removing the w3m thing

**Comment by coolo** [comment]
Package in case showing how many files are 'hidden' by forcing them into w3m:

```
"Mojolicious-8.17/t/mojolicious/ojo.t.w3m-render.txt"
"Mojolicious-8.17/lib/Mojolicious/resources/templates/mojo/exception.html.ep.w3m-render.txt",
"Mojolicious-8.17/lib/Mojolicious/Guides/Rendering.pod.w3m-render.txt",
"Mojolicious-8.17/t/mojolicious/restful_lite_app.t.w3m-render.txt",
"Mojolicious-8.17/t/mojolicious/rebased_lite_app.t.w3m-render.txt",
"Mojolicious-8.17/t/mojolicious/templates/layouts/green.html.epl.w3m-render.txt",
"Mojolicious-8.17/t/mojolicious/external_lite_app.t.w3m-render.txt",
"Mojolicious-8.17/examples/chat.pl.w3m-render.txt",
"Mojolicious-8.17/lib/Mojolicious/resources/templates/mojo/not_found.html.ep.w3m-render.txt",
"Mojolicious-8.17/lib/Mojolicious/resources/templates/mojo/debug.html.ep.w3m-render.txt"
"Mojolicious-8.17/t/mojolicious/templates/layouts/default.html.epl.w3m-render.txt",
"Mojolicious-8.17/README.md.w3m-render.txt",

**Comment by kraih** [comment]
Agreed about the HTML processing. Given our other problems with `File::Unpack` it would seem like a good idea to move as much code as possible out of it.

File: lib/Cavil/PostProcess.pm [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package Cavil::PostProcess;
+use Mojo::Base -base;
+
+has 'hash';
```
Comment (kraih) [inline lib/Cavil/PostProcess.pm]: This could probably use a more descriptive name.

File: lib/Cavil/PostProcess.pm [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+has max_line_length => 125;
+
+sub _split_find_a_good_spot {
+  my ($self, $line) = @_;
+
+  my $index  = $self->max_line_length;
+  my $length = length($line);
+  return $length if ($index > $length);
+  my %splits = (' ' => 1, ';' => 1, '{' => 1, '}' => 1, '"' => 0);
+  while ($index > $self->max_line_length * 0.7) {
```
Comment (kraih) [inline lib/Cavil/PostProcess.pm]: The repeated calls to `$self->max_line_length`  in this method are quite costly. Better assign the value to a variable in the beginning.

File: lib/Cavil/Checkout.pm [hunk 3]
```diff
@@ -148,9 +151,19 @@ sub unpack {
...
+    -f $log ? warn $err : die $err;
+    return;
+  }
+
+  my $unpacked  = decode_json($dir->child('.unpacked.json')->slurp);
+  my $processor = Cavil::PostProcess->new($unpacked);
+  $processor->postprocess;
+  open(my $fh, '>', $dir->child('.postprocessed.json'));
+  print $fh encode_json($processor->hash);
+  close($fh);
```
Comment (kraih) [inline lib/Cavil/Checkout.pm]: [Mojo::File::spurt](https://metacpan.org/pod/Mojo::File#spurt) is more efficient for writing a whole file at once. (`$dir->child('.postprocessed.json')->spurt(encode_json($processor->hash))`)

File: lib/Cavil/PostProcess.pm [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  }
+  return $to;
+}
+
+sub new { shift->SUPER::new(hash => shift) }
+
+sub postprocess {
+  my $self = shift;
+
+  for my $fname (keys %{$self->hash->{unpacked}}) {
```
Comment (kraih) [inline lib/Cavil/PostProcess.pm]: The repeated calls to `$self->hash` are very expensive again. Just put the value into a variable.

File: t/postprocess.t [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+use Mojo::JSON 'decode_json';
+use Cavil::Checkout;
+use Mojo::Util 'dumper';
+
+my $dir = path(__FILE__)->dirname->child('legal-bot');
+
+sub temp_copy {
+  my $from   = $dir->child(@_);
+  my $target = tempdir;
+  copy "$_", $target->child($_->basename) for $from->list->each;
```
Comment (kraih) [inline t/postprocess.t]: `Mojo::File` also has a [copy_to](https://metacpan.org/pod/Mojo::File#copy_to) method.

File: t/postprocess.t [hunk 6]
```diff
@@ -0,0 +1,62 @@
+use Mojo::Base -strict;
+
+use Test::More;
+use File::Copy 'copy';
+use Mojo::File qw(path tempdir);
+use Mojo::JSON 'decode_json';
+use Cavil::Checkout;
+use Mojo::Util 'dumper';
```
Comment (kraih) [inline t/postprocess.t]: Unused import.

File: t/checkout.t [hunk 7]
```diff
@@ -5,6 +5,7 @@ use File::Copy 'copy';
 use Mojo::File qw(path tempdir);
 use Mojo::JSON 'decode_json';
 use Cavil::Checkout;
+use Mojo::Util 'dumper';
```
Comment (kraih) [inline t/checkout.t]: Unused import.

Comment (kraih) [reply t/checkout.t]: If it's just for debug use, you can use `warn Mojo::Util::dumper ...;` without import anywhere in code that uses anything with Mojo in the name.

File: t/postprocess.t (hunk 5 above)
Comment (coolo) [reply t/postprocess.t]: I just copied it from checkout.t :)

**Comment by coolo** [comment]
It's processing for hours now one file:

/data/auto-co/legal-bot/ghc/b7310889c89b91183c89e01cee9d88d2/.unpacked/ghc-8.6.5/libraries/binary/generics-bench.cache

Sounds like I need to avoid copying large strings - aka substr()

**Comment by coolo** [comment]
I guess we're falling victim of 
https://stackoverflow.com/questions/25006875/performance-issue-with-substr-on-a-very-long-utf-8-string

File: lib/Cavil/PostProcess.pm (hunk 1 above)
Comment (coolo) [reply lib/Cavil/PostProcess.pm]: I agree, but can't come up with one :(

File: t/postprocess.t (hunk 5 above)
Comment (kraih) [reply t/postprocess.t]: Heh, i think that code might be why i added `copy_to` upstream then. 😆

**Comment by kraih** [comment]
Trying to be smart about where to break a line seems like a huge waste of resources to me anyway. There is just no way you can do better than break on whitespace with all the file formats we are dealing with.

**Comment by coolo** [comment]
Yeah, I came to the same conclusion. I wanted to be smart so it doesn't look like complete shit :(

**Pull request merged**